setup errors if CRAN version missing

in this ci job https://github.com/animint/animint2/actions/runs/18131727434/job/51599861370
there is an error
```
Run mkdir -p ~/.R
* installing *source* package ‘animint2.352f7e10040cb9de6ddd16416d342e9746c14c7a’ ...
** this is package ‘animint2.352f7e10040cb9de6ddd16416d342e9746c14c7a’ version ‘2025.9.16’
** using staged installation
** R
** data
*** moving datasets to lazyload DB
** inst
** byte-compile and prepare package for lazy loading
** help
Error : geom_bar.Rd:103: there is no package called ‘animint2’
ERROR: installing Rd objects failed for package ‘animint2.352f7e10040cb9de6ddd16416d342e9746c14c7a’
* removing ‘/home/runner/work/_temp/Library/animint2.352f7e10040cb9de6ddd16416d342e9746c14c7a’
Error in atime_versions_install(Package, normalizePath(pkg.path), new.Package.vec,  : 
  '/opt/R/4.5.1/lib/R/bin/R' CMD INSTALL -l '/home/runner/work/_temp/Library' '/tmp/Rtmpj0W34B/file33b049dcf455/animint2.352f7e10040cb9de6ddd16416d342e9746c14c7a' returned error status code 1
Calls: <Anonymous> ... system.time -> do.call -> <Anonymous> -> atime_versions_install
Timing stopped at: 9.509 1.014 9.867
Execution halted
Error: Process completed with exit code 1.
```
that Rd line is
```
  \Sexpr[results=rd,stage=build]{animint2:::rd_aesthetics("geom", "bar")}
```
this could be fixed via pkg.edit.fun (remove man files, or change the code to animint2.sha:::) but an even more user-friendly fix would be to install the un-modified package in atime_versions.

--- a/NEWS
+++ b/NEWS
@@ -1,6 +1,7 @@
 Changes in version 2026.5.29
 
 - atime_pkg() now uses only a common subset of columns of the measurements table from each test, so that each test can have different units to analyze.
+- atime_versions() runs setup using the HEAD commit (if present).
 
 Changes in version 2026.4.2
 

--- a/R/versions.R
+++ b/R/versions.R
@@ -132,8 +132,14 @@ atime_versions <- function(pkg.path, N=default_N(), setup, expr, sha.vec=NULL, t
   install.seconds <- system.time({
     ver.exprs <- do.call(atime_versions_exprs, ver.args)
   })[["elapsed"]]
+  sub.setup <- substitute(setup)
+  Package <- attr(ver.exprs, "Package")
+  if(is.character(Package)){
+    new.Package <- attr(ver.exprs, "new.Package")
+    sub.setup <- expr_pkg(sub.setup, Package, new.Package, check=FALSE)
+  }
   a.args <- list(
-    N, substitute(setup), ver.exprs, times, seconds.limit, verbose, result, N.env.parent)
+    N, sub.setup, ver.exprs, times, seconds.limit, verbose, result, N.env.parent)
   bench.seconds <- system.time({
     out.list <- do.call(atime, a.args)
   })[["elapsed"]]
@@ -157,7 +163,20 @@ get_sha_vec <- function(sha.vec, dots.vec){
     stop("each ... argument value and sha.vec element must be a string (package version, length=1, not NA), problems: ", paste(names(SHA.vec[is.problem]), collapse=", "))
   }
   SHA.vec
-}  
+}
+
+expr_pkg <- function(expr, Package, new.Package, check=FALSE){
+  old.lines <- capture.output(expr)
+  new.lines <- gsub(
+    paste0(Package,"::"),
+    paste0(new.Package,"::"),
+    old.lines,
+    fixed=TRUE)
+  if(check && Package!=new.Package && identical(old.lines,new.lines)){
+    stop(sprintf("expr=%s should contain at least one instance of %s:: to replace with %s::", old.lines, Package, new.Package))
+  }
+  str2lang(paste(new.lines, collapse="\n"))
+}
 
 atime_versions_exprs <- function(pkg.path, expr, sha.vec=NULL, verbose=FALSE, pkg.edit.fun=pkg.edit.default, ...){
   formal.names <- names(formals())
@@ -175,20 +194,16 @@ atime_versions_exprs <- function(pkg.path, expr, sha.vec=NULL, verbose=FALSE, pk
     ifelse(SHA.vec=="", "", "."), 
     SHA.vec)
   a.args <- list()
+  sub.expr <- substitute(expr)
   for(commit.i in seq_along(SHA.vec)){
     sha <- SHA.vec[[commit.i]]
     commit.name <- names(SHA.vec)[[commit.i]]
     new.Package <- new.Package.vec[[commit.i]]
-    old.lines <- capture.output(substitute(expr))
-    new.lines <- gsub(
-      paste0(Package,"::"),
-      paste0(new.Package,"::"),
-      old.lines,
-      fixed=TRUE)
-    if(Package!=new.Package && identical(old.lines,new.lines)){
-      stop(sprintf("expr should contain at least one instance of %s:: to replace with %s::", Package, new.Package))
+    if(grepl("HEAD", commit.name)){
+      attr(a.args, "Package") <- Package
+      attr(a.args, "new.Package") <- new.Package
     }
-    a.args[[commit.name]] <- str2lang(paste(new.lines, collapse="\n"))
+    a.args[[commit.name]] <- expr_pkg(sub.expr, Package, new.Package, check=TRUE)
     atime_versions_install(
       Package, normalizePath(pkg.path),
       new.Package.vec, SHA.vec, verbose, pkg.edit.fun)

--- a/R/versions.R
+++ b/R/versions.R
@@ -132,11 +132,15 @@ atime_versions <- function(pkg.path, N=default_N(), setup, expr, sha.vec=NULL, t
   install.seconds <- system.time({
     ver.exprs <- do.call(atime_versions_exprs, ver.args)
   })[["elapsed"]]
-  sub.setup <- substitute(setup)
-  Package <- attr(ver.exprs, "Package")
-  if(is.character(Package)){
-    new.Package <- attr(ver.exprs, "new.Package")
-    sub.setup <- expr_pkg(sub.setup, Package, new.Package, check=FALSE)
+  if(missing(setup)){
+    sub.setup <- quote({})
+  }else{
+    sub.setup <- substitute(setup)
+    Package <- attr(ver.exprs, "Package")
+    if(is.character(Package)){
+      new.Package <- attr(ver.exprs, "new.Package")
+      sub.setup <- expr_pkg(sub.setup, Package, new.Package, check=FALSE)
+    }
   }
   a.args <- list(
     N, sub.setup, ver.exprs, times, seconds.limit, verbose, result, N.env.parent)

--- a/man/atime_versions.Rd
+++ b/man/atime_versions.Rd
@@ -16,8 +16,10 @@ atime_versions(
   \item{N}{numeric vector of data sizes to vary.}
   \item{setup}{
     expression to evaluate for every data size, before timings.
-    In contrast to \code{expr},
-    no replacement of \code{Package::} is performed.
+    If one of the versions contains name HEAD, then
+    \code{Package::} is replaced with
+    \code{Package.SHA_for_HEAD::}
+    (in other words, setup is run using the HEAD commit).
   }
   \item{expr}{
     code with package double-colon prefix, for example

--- a/tests/testthat/test-versions.R
+++ b/tests/testthat/test-versions.R
@@ -34,7 +34,7 @@ if(requireNamespace("ggplot2"))test_that("atime_pkg produces tests_all_facet.png
   ## https://github.com/tdhock/binsegRcpp/tree/atime-test-funs
   atime.dir <- file.path(tdir, ".ci", "atime")
   unlink(file.path(atime.dir, "*"))
-  git2r::checkout(repo, branch="atime-test-funs", force=TRUE)
+  git2r::checkout(repo, branch="test-setup-HEAD", force=TRUE)
   options(repos="http://cloud.r-project.org")#required to check CRAN version.
   plist <- atime::atime_pkg(tdir, ".ci")
   tests.RData <- file.path(atime.dir, "tests.RData")

--- a/tests/testthat/test-versions.R
+++ b/tests/testthat/test-versions.R
@@ -26,7 +26,7 @@ test_that("atime_versions_exprs error when expr does not contain pkg:", {
       "Before"="be2f72e6f5c90622fe72e1c315ca05769a9dc854",
       "Regression"="e793f53466d99f86e70fc2611b708ae8c601a451", 
       "Fixed"="58409197426ced4714af842650b0cc3b9e2cb842") 
-  }, "expr should contain at least one instance of binsegRcpp:: to replace with binsegRcpp.be2f72e6f5c90622fe72e1c315ca05769a9dc854:", fixed=TRUE)
+  }, "expr=dt[, .(vs = (sum(val))), by = .(id)] should contain at least one instance of binsegRcpp:: to replace with binsegRcpp.be2f72e6f5c90622fe72e1c315ca05769a9dc854::", fixed=TRUE)
 })
 
 if(requireNamespace("ggplot2"))test_that("atime_pkg produces tests_all_facet.png and tests_preview_facet.png on atime-test-funs", {

